feat: enhance bootloader for multiple flash support

badrbouslikhin opened the pull request:
## Summary
This pull request introduces a enhancement to the `BootLoaderConfig` and `FirmwareUpdaterConfig` in the `embassy-boot` crate. The primary goal is to implement support for multiple flash partitions, enabling the bootloader and firmware updater to handle `ActiveFlash`, `DFUFlash`, and `StateFlash` independently. This change is particularly beneficial for systems with dual-bank or external flash setups.

## Key Changes
- **Configurations Refactored**: Both `BootLoaderConfig` and `FirmwareUpdaterConfig` have been updated to support different flash partition types.
- **Partition Handling**: Adjustments made in various modules for alignment with the new multi-flash support, specifically in the instantiation of `BlockingPartition` and `Partition`.
- `bootloader/stm32` and `bootloader/stm32wb-dfu` have been updated and a new example `bootloader/stm32-dual-bank` has been introduced.

## Testing
- The changes have been tested in the blocking flavor on an STM32H747XI.

## TODO

- [x] Update all examples
- [x] Test on other targets (is the CI capable of performing this?)
- [x] Improve `from_linkerfile_blocking` api?

Your review would be greatly appreciated!

Specifically, I am not fully satisfied with the `from_linkerfile_blocking` when the same flash is used:
```rust
let config = BootLoaderConfig::from_linkerfile_blocking(&flash, &flash, &flash);
```
I see two ways to handle this: a builder pattern or wrapping `&flash` with an Option.

badrbouslikhin commented:
Ah, that explains why the Embassy bootloader doc. page says it `the bootloader supports both internal and external flash`!

Once I get your thoughts on the changes and `from_linkerfile_blocking` api, I'll proceed with the rest :).

lulf commented on embassy-boot/src/boot_loader.rs:
```diff
@@ -49,16 +49,20 @@ pub struct BootLoaderConfig<ACTIVE, DFU, STATE> {
 }
 
-impl<'a, FLASH: NorFlash>
+impl<'a, ActiveFlash: NorFlash, DFUFlash: NorFlash, StateFlash: NorFlash>
```
Could you change the names ACTIVE, DFU and STATE which is also used elsewhere in the code.

badrbouslikhin replied:
Done in af2b4df.

lulf commented on embassy-boot/src/boot_loader.rs:
```diff
@@ -49,16 +49,20 @@ pub struct BootLoaderConfig<ACTIVE, DFU, STATE> {
     /// Create a bootloader config from the flash and address symbols defined in the linkerfile
     // #[cfg(target_os = "none")]
-    pub fn from_linkerfile_blocking(flash: &'a Mutex<NoopRawMutex, RefCell<FLASH>>) -> Self {
+    pub fn from_linkerfile_blocking(
```
Personally I don't think this API is too bad. Using a builder doesn't provide that much gain, but perhaps a better rustdoc order explaining the parameters would be appropriate

badrbouslikhin replied:
Done in 634c409.

badrbouslikhin commented:
I updated and compiled all examples. 
Some of them required changes in the `memory.x` most likely due to recent code inflation (they were not compiling before this change). I adjusted linker files but I do not have the boards to validate the change.

Side note: is it useful to maintain that many `boot/application` STM32 examples? They look very similar.

I am not sure how to fix CI failures though.

lulf commented:
There is no HIL test for the bootloader, but any examples should be added to the ci.sh to ensure they always compile. It's unfortunate that the size increased previously though, and it goes undetected because in ci it compiles with --features skip-include to avoid the double pass.

The ci/test is failing on the unit tests in the bootloader though https://ci.embassy.dev/jobs/d45560625688

badrbouslikhin commented:
Alright, CI is happy now. Let me know if there's anything else I need to do.

lulf reviewed (approved):
LGTM, thanks!
